### ISSUE TITLE
[Pocketbook Lux 3] KOReader displays the book cover when exiting the application

### ISSUE THREAD

**Issue by NicolasPetton** [koreader/koreader#3460]
* KOReader version: koreader-nightly-20171004
* Device: Pocketbook touch lux 3

#### Issue

When exiting koreader (pressing the home button on the device), KOReader displays the book cover on top of the Pocketbook home screen (I can click on the icons behind the image though, and any action that triggers a refresh of the display fixes the problem).

#### Steps to reproduce

1. Start KOReader
2. Open a book
3. Exit KOReader (press the Home button)

**Comment by NicolasPetton** [comment]
Any news on this issue? Do you need more info to be able to reproduce it?

**Comment by Frenzie** [comment]
Pinging @mk-f 

Most of us don't have a PocketBook, so info isn't exactly the problem. ;-)

At a guess there might be some confusion between exiting KOReader and suspending or shutting down KOReader in your specific example, in the sense that I don't think you'd expect KOReader to draw anything when exiting. But the bigger question might be whether solving that minor issue wouldn't just result in the same problem, except with the display remaining as it is?

**Comment by NicolasPetton** [comment]
Frans de Jonge <notifications@github.com> writes:

> At a guess there might be some confusion between exiting KOReader and
> suspending or shutting down KOReader in your specific example, in the
> sense that I don't think you'd expect KOReader to draw anything when
> exiting. But the bigger question might be whether solving that minor
> issue wouldn't just result in the same problem, except with the
> display remaining as it is?

It indeed looks like KOReader gets suspended when pushing the home
button on the device.

I wouldn't consider this minor though, as pushing the home button on the
Pocketbook makes the device unusable unless the user understands that
he/she needs to trigger full screen refresh somehow (a long press on the
home button opens the task manager, which forces a full refresh of the
screen).

When exiting KOReader from KOReader's menu, I correctly get back to the
home screen of the pocketbook though.  Would it make sense to exit
KOReader when pushing this button?

**Comment by Frenzie** [comment]
> When exiting KOReader from KOReader's menu, I correctly get back to the home screen of the pocketbook though. Would it make sense to exit KOReader when pushing this button?

Only someone with a PocketBook can answer that, but from what you've said I would assume so.

It should be simple enough. There's the home key for PocketBook:

https://github.com/koreader/koreader/blob/a39db783c0e79557324ea9c142a8431b040a95fc/frontend/device/pocketbook/device.lua#L31

And there's the generic exit function I wrote for PC. It should be refactored in that case; I suppose it also be lazily extended to the relevant form of `or KEY_HOME` for a quick fix. Something like `[KEY_HOME] = "Exit"` in the device definition and `or keycode == "Power"` for the check. (Just brainstorming.)

https://github.com/koreader/koreader/blob/a39db783c0e79557324ea9c142a8431b040a95fc/frontend/device/input.lua#L302-L326

**Comment by NicolasPetton** [comment]
Frans de Jonge <notifications@github.com> writes:

> Only someone with a PocketBook can answer that, but from what you've
> said I would assume so.

I guess so as well, pressing the home button exits other applications.

**Comment by mk-f** [comment]
My Device does not have a Home-Button ;). 

@NicolasPetton can you test with [this](https://github.com/mk-f/koreader/blob/f27891eb1502b8247e17d604d05a4d22154cfc4f/frontend/device/pocketbook/device.lua) quick and dirty? 
@Frenzie Any objections on broadcasting "Close" instead of using the Exit-function? Would not require touching the generic ``device/input.lua`` and the "Close"-broadcast is already used in ``pocketbook/device.lua``

**Comment by Frenzie** [comment]
@mk-f 
Yes, unfortunately, but I'd still merge your commit. ;-) Explanation follows.

Closing all widgets may result in a failure to close or unsaved settings. See https://github.com/koreader/koreader/commit/a39db783c0e79557324ea9c142a8431b040a95fc

```diff
+                -- if we don't do this you can get a situation where either the
+                -- program won't exit due to remaining widgets until they're
+                -- dismissed or if the callback forces all widgets to close,
+                -- that the save document ConfirmBox is also closed
```

Either the entire thing should be refactored so that Close actually properly takes care of everything, or what seemed architecturally more straightforward to me at the time was to separately take care of some special cases.

tl;dr https://github.com/koreader/koreader/commit/bc9f60034ee7aff82be20f23c371ea5d0be087cc (which I hadn't noticed) should already be problematic in some edge cases today.

But! Your commit is perfectly acceptable to me since it's not making any problems any worse. It's really a separate issue.

**Comment by pazos** [comment]
Duplicate of #5140

Screensaver should be disabled on pocketbooks for the moment. It just introduces random noise on a platform we barely support :/
